oauth/src/OAuth/OAuth2.hs

```diff
@@ -1,32 +1,56 @@
+{-# OPTIONS -Werror #-}
 module OAuth.OAuth2
-  ( AuthCode(..)
-  , AuthenticationResponse(..)
+  ( AuthCode
+  , AuthorizationRequest(..)
   , Config(..)
   , Env
-  , authenticate
-  , authorizationRequestUrl
+  , RedirectURI
+  , State
+  , TokenResponse(..)
+  , authorizationRequestURI
+  , getAuthorizationRequest
+  , getTokenRequest
+  , requestToken
   )
 where
 
+import Control.Monad (replicateM)
 import Data.Bifunctor (second)
+import Data.Bounded
 import Data.ByteString (ByteString)
+import Data.Word (Word8)
 import OAuth.Prelude
 import Prelude(Integer)
 
-import qualified Crypto.Nonce               as Nonce
-import qualified Data.Aeson                 as JSON
-import qualified Data.Text                  as Text
-import qualified Data.Text.Encoding         as Text
-import qualified Network.HTTP.Client        as HTTP
-import qualified Network.HTTP.MClient       as HTTP
-import qualified Network.HTTP.Types         as HTTP
-import qualified Network.HTTP.Types.URI     as URI
+import qualified Data.Aeson             as JSON
+import qualified Data.ByteString        as BS
+import qualified Data.ByteString.Base64 as Base64
+import qualified Data.Text              as Text
+import qualified Data.Text.Encoding     as Text
+import qualified Network.HTTP.Client    as HTTP
+import qualified Network.HTTP.MClient   as HTTP
+import qualified Network.HTTP.Types     as HTTP
+import qualified Network.HTTP.Types.URI as URI
+import qualified System.Random          as System
+
+type AccessType            = BoundText "OAuth2 AccessType"
+type AuthCode              = BoundText "OAuth2 AuthCode"
+type AuthorizationEndpoint = BoundText "OAuth2 AuthorizationEndpoint"
+type ClientId              = BoundText "OAuth2 ClientId"
+type ClientSecret          = BoundText "OAuth2 ClientSecret"
+type GrantType             = BoundText "OAuth2 GrantType"
+type Nonce                 = BoundText "OAuth2 Nonce"
+type RedirectURI           = BoundText "OAuth2 RedirectURI"
+type ResponseType          = BoundText "OAuth2 ResponseType"
+type Scope                 = BoundText "OAuth2 Scope"
+type State                 = BoundText "OAuth2 State"
+type TokenEndpoint         = BoundText "OAuth2 TokenEndpoint"
 
 data Config = Config
-  { authURL      :: Text
-  , clientId     :: Text
-  , clientSecret :: Text
-  , tokenURL     :: Text
+  { authorizationEndpoint :: AuthorizationEndpoint
```

Comment:
Comment: I suppose in future we may have a proper URL/endpoint type that takes into account URL properties.